src/Spago/Watch.hs

```diff
@@ -47,24 +53,43 @@ fileWatchConf
   -> ((Set.Set Glob.Pattern -> m ()) -> m ())
   -> m ()
 fileWatchConf watchConfig shouldClear inner = withManagerConf watchConfig $ \manager -> do
-    allGlobs <- liftIO $ newTVarIO Set.empty
-    dirtyVar <- liftIO $ newTVarIO True
-    watchVar <- liftIO $ newTVarIO Map.empty
+    allGlobs  <- liftIO $ newTVarIO Set.empty
+    dirtyVar  <- liftIO $ newTVarIO True
+    -- `lastEvent` is used for event debouncing.
+    -- We don't use built-in debouncing because it does not work well with some
+    -- text editors (#346).
+    lastEvent <- liftIO $ do
+      timeNow <- getCurrentTime
+      newTVarIO (timeNow, "")
+    watchVar  <- liftIO $ newTVarIO Map.empty
 
     let redisplay maybeMsg = do
           when (shouldClear == DoClear) $ liftIO clearScreen
           mapM_ echoStr maybeMsg
 
     let onChange event = do
-          globsUnsafe <- liftIO $ readTVarIO allGlobs
-          let shouldRebuild globs = or $ fmap (\glob -> Glob.match glob $ Watch.eventPath event) $ Set.toList globs
-          when (shouldRebuild globsUnsafe) $ do
-            redisplay $ Just $ "File changed, rebuilding: " <> show (Watch.eventPath event)
-          liftIO $ atomically $ do
-            globs <- readTVar allGlobs
-            when (shouldRebuild globs)
+          timeNow <- liftIO getCurrentTime
+
+          rebuilding <- liftIO $ atomically $ do
+            globs                <- readTVar allGlobs
+            (lastTime, lastPath) <- readTVar lastEvent
+
+            let shouldRebuild =
```

Comment:
I find these 4 lines of logic a bit convoluted to parse, but I feel we cannot get them any simpler. Could you add some comments here on top with the rules to decide if we should rebuild? (basically what the logic implements but in natural language)